neet/sensitivity.py

```diff
@@ -35,33 +39,178 @@ def sensitivity(net, state):
         raise(TypeError("net must be a boolean network"))
 
     # list Hamming neighbors
-    neighbors = hamming_neighbors(state)
+    neighbors = _hamming_neighbors(state)
 
     nextState = net.update(state)
 
     # count sum of differences found in neighbors of the original
     s = 0.
     for neighbor in neighbors:
-        newState = net.update(neighbor)
-        s += boolean_distance(newState, nextState)
+        if transitions is not None:
+            newState = transitions[_fast_encode(neighbor)]
+        else:
+            newState = net.update(neighbor)
+        s += _boolean_distance(newState, nextState)
 
     return s / net.size
 
-def boolean_distance(state1,state2):
+def difference_matrix(net, state, transitions=None):
+    """
+    Returns matrix answering the question:
+    Starting at the given state, does flipping the state of node j
+    change the state of node i?
+    """
+    # set up empty matrix
+    N = len(state)
+    Q = np.empty((N,N))
+    
+    # list Hamming neighbors (in order!)
+    neighbors = _hamming_neighbors(state)
+
+    nextState = net.update(state)
+
+    # count differences found in neighbors of the original
+    s = 0.
+    for j,neighbor in enumerate(neighbors):
+        if transitions is not None:
+            newState = transitions[_fast_encode(neighbor)]
+        else:
+            newState = net.update(neighbor)
+        Q[:,j] = [ ( nextState[i] + newState[i] )%2 for i in range(N) ]
+        
+    return Q
+
+def _states_limited(nodes,state0):
+    """
+    All possible states that vary only nodes with given indices.
+    """
+    if len(nodes) == 0:
+        return [state0]
+    for i in nodes:
+        stateFlipped = copy.copy(state0)
+        stateFlipped[nodes[0]] = (stateFlipped[nodes[0]]+1)%2
+        return _states_limited(nodes[1:],state0) + _states_limited(nodes[1:],stateFlipped)
```

Comment:
I'm a huge fan of recursion, but Python is shockingly bad at it. Implementing `_states_limited` iteratively would bring further performance improvements (not that they're really necessary).

Reply:
Fair enough, but I think that it's fast enough now and that it's late enough on a Friday afternoon for me to leave this for later.  :)